### ISSUE TITLE
add --chromaloc 2 to x265 parameters automatically if source is 4.2:0 (Type 2)

### ISSUE THREAD

**Issue by E-t-z** [cdgriffith/FastFlix#190]
A bit unsure on this, if it is even required, some sources say it is definitely required if you want to retain HDR, others say it is optional.

Maybe @cdgriffith can comment on this (you wrote terrific article on HDR encoding after all) and if needed, could implement it?

**Comment by cdgriffith** [comment]
Never heard of that before actually. Know of any example sources that have type 2 chroma? Would need to find out how that is displayed via `ffprobe`.

If you happen to have one, would appreciate output of:

```
ffprobe -v quiet -loglevel panic -print_format json -show_format -show_streams <file>
```

**Comment by E-t-z** [comment]
Most UHD BR's with HDR should have it, but I have some sample clips myself as well.

Type 2 does not have anything chromaloc related present in JSON.

But, when you don't specify it and sue Defaults, this row appears on HEVC stream.
`"chroma_location": "topleft",`

In mediainfo, it shows like  this:

Type 2:
![image](https://user-images.githubusercontent.com/24243843/105960973-9795e780-6086-11eb-856e-02f128223e65.png)

Default (Without chromaloc specified):
![image](https://user-images.githubusercontent.com/24243843/105961063-b3998900-6086-11eb-83e8-6a682536f40c.png)

**Comment by E-t-z** [comment]
Some documentation  https://www.itu.int/rec/dologin_pub.asp?lang=e&id=T-REC-H.Sup19-201910-I!!PDF-E&type=items

![image](https://user-images.githubusercontent.com/24243843/105961904-cc566e80-6087-11eb-8aa8-b57f46001eed.png)

This table says, that BT.2020 specification requires Type 2.

**Comment by E-t-z** [comment]
As this seems to be quite confusing, my proposal would be to use it only, when present in source material.
It kind of assures that everything is preserved as it should.

**Comment by E-t-z** [comment]
My current assumption is, that it should be always used, when dealing with WCG/HDR content and YUV.

**Comment by cdgriffith** [comment]
Thanks for the details! 

Luckily it does display in ffprobe:  `"chroma_location": "left"` looks to be 0 and `"chroma_location": "topleft"` looks to be 2.  

And  you are correct that it seems a movies encoded with FastFlix without that does switch it from `topleft` to `left`. (There seems to be debate if anything actually uses that flag right now, is possibly why no one has yet noticed https://forum.doom9.org/archive/index.php/t-174491.html) 

https://x265.readthedocs.io/en/3.1/cli.html says to refer to the HEVC specification to find what the 5 possible options mean, did a quick search through https://www.itu.int/rec/T-REC-H.265-201911-I/en without much luck, would like to get them all defined for future proofing.

**Comment by E-t-z** [comment]
For me Type 2, does not seem to reflect anything in ffprobe output, at least BR's which I checked have nothing in that output.
That element is simply not present...but they are Detected as 4:2:0 (Type 2).

You are probably right, that nothing uses it at the moment, as my AndroidTV does not seem to care.
Still, it would be nice to follow the standards, in case something does and messes up the colors 😆

**Comment by E-t-z** [comment]
@cdgriffith BTW all HEVC possibel chroma locations are described here:
https://www.itu.int/rec/dologin_pub.asp?lang=e&id=T-REC-H.265-201612-S!!PDF-E&type=items

On page 384

TL;DR: They are basically: left, center, top_left, top, bottom_left, bottom

**Comment by cdgriffith** [comment]
> I think best way to understand, how it is actually reflected in ffprobe, would be to encode 6 sample files, with chromaloc flag 0...5 and one without that flag and then compare outputs.

You know, I felt so smart going and brute forcing it just like that, and was coming back to report the results and you already posted the idea 😭


 | chromaloc | ffmpeg chroma_location |
| --- | --- | 
| 0 | left |
| 1 | center |
| 2 | topleft |
| 3 | top |
| 4 | bottomleft |
| 5 | bottom |

**Comment by E-t-z** [comment]
And now, checkbox would be nice "Preserve Chrome location", checked by default for HDR 😄  

Too bad, I suck on python...maybe you could rewrite it in C# 😋

**Comment by cdgriffith** [comment]
I think it's safe to just always insert it if the incoming source provides a chroma_location. 

My only wonder now is if doing the remove_hdr process changes the chroma_location..hmm..

**Comment by E-t-z** [comment]
That is an interesting question indeed.
But if you look this table BT.709 has Chroma_location=0 always.

AFAIK, you cannot remove HDR and keep BT.2020 😉

**Comment by cdgriffith** [comment]
Currently it does take it down to bt709, so hopefully just always setting / leaving it at 0 is proper for that. 

Filter used: 
```
zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt709,tonemap=tonemap={tone_map}:desat=0,zscale=t=bt709:m=bt709:r=tv,format=yuv420p"
```

**Comment by E-t-z** [comment]
As 0 seems to be default as well, I don't think it would cause an issue.

**Comment by cdgriffith** [comment]
I hope not, and just did some test encodes with Handbrake to see how they handled it, they seem to ignore it entirely as well (everything comes out 0).

I'm still planning on adding the passthrough, just glad I'm not the only one who missed the memo 😉

**Comment by E-t-z** [comment]
As I stated at the beginning, I'm quite unsure how much it actually matters in real life, but it seemed to violate the spec, 
so I simply started digging.

Even checked couple of encodes (of UHD BR's I own) on torrent sites, to see how famous scene groups handle it.
And they all seemed to preserve "chromaloc" for some reason...and you already know the rest I guess... 😄

**Comment by E-t-z** [comment]
Anyway, I think we can close this now.

**Comment by cdgriffith** [comment]
I just like to keep these open until actually released to keep track of progress, thanks again for all the info!

**Comment by cdgriffith** [comment]
Fixed in 4.2.0 https://github.com/cdgriffith/FastFlix/releases/tag/4.2.0

**Comment by ValZapod** [comment]
> that it should be always used, when dealing with WCG/HDR content and YUV.

No, only if it is not Dolby Vision profile 5 that just uses 0, left. In all other cases yes.

> on torrent sites, to see how famous scene groups handle it.
> And they all seemed to preserve "chromaloc" for some reason...and you already know the rest I guess...

Well, of course they do. After all most of them just webdl or rip already encoded by Netfliz and Disney+ stuff!

**Comment by ValZapod** [comment]
> I hope not, and just did some test encodes with Handbrake to see how they handled it, they seem to ignore it entirely as well (everything comes out 0).
> 
> I'm still planning on adding the passthrough, just glad I'm not the only one who missed the memo wink

They do not handle it. It is a bug which was recently fixed in ffmpeg.

> it down to bt709, so hopefully just always setting / leaving it at 0 is proper for that.

No, it is not. When you convert to BT.709 you need to scale planes to different chroma sample location.

**Comment by cdgriffith** [comment]
> When you convert to BT.709 you need to scale planes to different chroma sample location with -vf scale. @ValZapod 

I don't fully follow, do you have an example?

**Comment by ValZapod** [comment]
> don't fully follow, do you have an example?

Imagine JPEG screenshot. You go either from Y'CbCr left chromaloc to R'G'B' and use RGB jpeg or you do 4:4:4 YCbCr and all is good **or you need to convert** left Y'CbCr to center Y'CbCr as used in jpeg. And vice versa. Same about HDR. All commands are here (my comments are under username Balling): https://forum.doom9.org/showthread.php?t=183145

Something like for ffmpeg -vf scale=out_color_matrix=bt2020nc:out_h_chr_pos=0:out_v_chr_pos=0
